minetest/minetest
# Add no_texture.png as fallback for unspecified textures

This adds `no_texture.png`, a new fallback texture for nodes, items and entities with unspecified textures.
The rationale behind this is basically to make error conditions in games and mods more visible and clearer and to help making texture bugs in your game more obvious.

Here's the current situation that I want to improve:

Currently, Minetest abuses `unknown_node.png` and `unknown_object.png` as fallback texture when a texture is missing. This can be misleading during development.

Moreover, if a non-node item has no `inventory_image`, Minetest behaves buggy: Its inventory icon is invisible and renders as a buggy white cube on your wieldhand. This is because that case isn't currently captured at all in the code.

Finally, I found a segfault in Minetest! If you register the following node and try to look at it in an inventory:

```
minetest.register_node("segfault:node_of_doom", {
drawtype="plantlike_rooted",
})
```
Minetest will segfault. This happens because there are no checks in place for `special_tiles`, which the rendering code of `plantlike_rooted` needs.

## What this PR does

* Add `no_texture.png`, a bright red 16×16 texture with "no texture" written on it
* In case of unspecified node textures, renders `no_texture.png` (was `unknown_node.png` before)
* In case of unspecified entity textures, renders `no_texture.png` (was `unknown_object.png` before)
* In case of unspecified item image of a non-node item, renders `no_texture.png` (was invisible/buggy before)
* Add the mod `broken` to DevTest which contains 1 node, 1 tool, 1 craftitem and 1 entity, all with empty definitions. Intended to test Minetest's fallback behavior of these things
* Fix a segfault when the game contains a node with this definition: `{drawtype="plantlike_rooted"}`

## What this PR does NOT do

* Change the "dummy image" (random color) when a texture is *missing* (i.e. no PNG file). This PR only cares about *unspecified* textures, not *missing* textures
* Remove `unknown_node.png` or `unknown_object.png`. Those file are still used, but now exclusively for unknown nodes and objects only
* Add error messages (I'm not sure where and when to put them)

## Q&A

Q: "Why did you remove the node name in the unknown node infotext?"
A: It turned out, the client (no longer?) knows about the true name of unknown nodes, instead only gets "unknown" when it ask about the node name. Since that is not useful information to the user/debugger, I removed that.

Q: "Why did you set `backface_culling=false` in `src/nodedef.cpp`?"
A: Because I want to make sure the `no_texture.png` is always visible no matter from which side you look at it. Some drawtypes have backface culling enabled by default. Since this is for developers anyway, performance is not an issue.

## To do

This PR is ready for review.

HINT: Review the 1st commit for better readabiltiy because the 2nd commit only touches indentation.

## How to test

Go to DevTest and check out the nodes of the mod `broken`. All of these should always render with the texture `no_texture.png`.
Use the entity spawner to spawn `broken:entity_with_empty_definition`. It should also render with `no_texture.png`.

Also, somehow get legitimate unknown nodes, objects and items into your world (maybe with a dummy mod that adds a node and item and entity, put these things into your world, then you exit the world, disable the dummy mod and enter the world again) and test if they still work as expected (infotext, unknown node/object texture).

Finally, test that everything else still works like before.
